### PR TITLE
DataTableBuilder actions column

### PR DESCRIPTION
**Pull request by florianlacreuse**

### CODE REVIEW

File: basic-application/basic-application-webapp/src/main/java/fr/openwide/core/basicapp/web/application/administration/component/UserGroupMembersPanel.java [hunk 1]
```diff
@@ -66,15 +68,56 @@ public UserGroupMembersPanel(String id, IModel<UserGroup> userGroupModel) {
...
+							.withLink(LinkFactory.userDescriptionLinkGeneratorFactory())
+							.withClass("text text-md")
+					.addActionColumn()
+							.addConfirmAction(
+									new BootstrapRenderer<User>() {
+										private static final long serialVersionUID = 1L;
+										@Override
+										protected BootstrapRendererInformation doRender(User value, Locale locale) {
+											return BootstrapRendererInformation.builder()
+													.tooltip(getString("administration.usergroup.members.delete", locale))
```
Comment (yrodiere) [inline basic-application/basic-application-webapp/src/main/java/fr/openwide/core/basicapp/web/application/administration/component/UserGroupMembersPanel.java]: In order to shorten the code for this use case, maybe ActionRenderers could include some parameterized versions of `remove()`, `trash()` and so on, where we would only provide the resource key? Like this:

```
    public static <T> BootstrapRenderer<T> remove(final String resourceKey) {
        return new BootstrapRenderer<T>() {
            private static final long serialVersionUID = 1L;
            @Override
            protected BootstrapRendererInformation doRender(T value, Locale locale) {
                return BootstrapRendererInformation.builder()
                        .label(getString(resourceKey, locale))
                        .tooltip(getString(resourceKey, locale))
                        .icon("fa fa-fw fa-trash-o")
                        .color(BootstrapColor.DANGER)
                        .build();
            }
        };
    }
```

Then this code would be much shorter.

File: basic-application/basic-application-webapp/src/main/java/fr/openwide/core/basicapp/web/application/common/renderer/ActionRenderers.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+						.label(getString("common.action.add", locale))
+						.tooltip(getString("common.action.add", locale))
+						.icon("fa fa-fw fa-plus")
+						.color(BootstrapColor.PRIMARY)
+						.build();
+			}
+		};
+	}
+
+	public static <T> BootstrapRenderer<T> remove() {
```
Comment (yrodiere) [inline basic-application/basic-application-webapp/src/main/java/fr/openwide/core/basicapp/web/application/common/renderer/ActionRenderers.java]: I'd be in favor of distinguishing between:
- `delete`, with an `fa-trash` icon and a "Delete" tooltip ("Supprimer" in French): really erase all information about this object, thus the "trash" icon.
- `remove`, with an `fa-times` icon and a "Remove" tooltip ("Retirer" in french): put an object out of somewhere, for instance out of a collection, thus the "times" icon, which is, in my opinion, a bit less harsh.

Of course it wouldn't be a rule for every project, but just a good practice for newly started project. Which is the purpose of the basic application, after all.

@florianlacreuse , @gsmet , what do you think?

File: basic-application/basic-application-webapp/src/main/java/fr/openwide/core/basicapp/web/application/administration/component/UserGroupMembersPanel.java (hunk 1 above)
Comment (yrodiere) [inline basic-application/basic-application-webapp/src/main/java/fr/openwide/core/basicapp/web/application/administration/component/UserGroupMembersPanel.java]: Actually, we could even provide a fully parametrized version, like this:

```
    public static <T> BootstrapRenderer<T> constant(final String resourceKey, final String icon, final IBootstrapColor color) {
        return new BootstrapRenderer<T>() {
            private static final long serialVersionUID = 1L;
            @Override
            protected BootstrapRendererInformation doRender(T value, Locale locale) {
                return BootstrapRendererInformation.builder()
                        .label(getString(resourceKey, locale))
                        .tooltip(getString(resourceKey, locale))
                        .icon(icon)
                        .color(color)
                        .build();
            }
        };
    }
```

File: owsi-core/owsi-core-components/owsi-core-component-wicket-more/src/main/java/fr/openwide/core/wicket/more/markup/html/repeater/data/table/builder/action/CoreActionColumn.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+public class CoreActionColumn<T, S extends ISort<?>> extends AbstractCoreColumn<T, S> {
+
+	private static final long serialVersionUID = -8748416651399483779L;
+
+	private final List<AbstractActionColumnElementFactory<T, ?>> factories = Lists.newArrayList();
+
+	public CoreActionColumn(IModel<String> headerLabelModel, List<AbstractActionColumnElementFactory<T, ?>> factories) {
+		super(headerLabelModel);
+		CollectionUtils.replaceAll(this.factories, factories);
```
Comment (yrodiere) [inline owsi-core/owsi-core-components/owsi-core-component-wicket-more/src/main/java/fr/openwide/core/wicket/more/markup/html/repeater/data/table/builder/action/CoreActionColumn.java]: @florianlacreuse `this.factories.addAll(factories)` would be more efficient here (one less list instantiation). If this snippet was copied from somewhere, could you please point it to me?

File: basic-application/basic-application-webapp/src/main/java/fr/openwide/core/basicapp/web/application/common/renderer/ActionRenderers.java (hunk 2 above)
Comment (gsmet) [inline basic-application/basic-application-webapp/src/main/java/fr/openwide/core/basicapp/web/application/common/renderer/ActionRenderers.java]: +1

File: owsi-core/owsi-core-components/owsi-core-component-wicket-more/src/main/java/fr/openwide/core/wicket/more/markup/html/repeater/data/table/builder/action/factory/AbstractActionColumnElementFactory.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+import fr.openwide.core.wicket.more.util.model.Detachables;
+
+
+public abstract class AbstractActionColumnElementFactory<T, F extends AbstractActionColumnElementFactory<T, ?>> implements IOneParameterComponentFactory<AbstractLink, IModel<T>> {
+
+	private static final long serialVersionUID = 8791565179874571105L;
+
+	private final BootstrapRenderer<? super T> renderer;
+
+	private Condition showLabelCondition = Condition.alwaysTrue();
```
Comment (yrodiere) [inline owsi-core/owsi-core-components/owsi-core-component-wicket-more/src/main/java/fr/openwide/core/wicket/more/markup/html/repeater/data/table/builder/action/factory/AbstractActionColumnElementFactory.java]: Shouldn't this be false by default? I find it very rare to display a button's label inside a table.

Comment (gsmet) [inline owsi-core/owsi-core-components/owsi-core-component-wicket-more/src/main/java/fr/openwide/core/wicket/more/markup/html/repeater/data/table/builder/action/factory/AbstractActionColumnElementFactory.java]: `import com.impossibl.postgres.utils.guava.Joiner`

-> once more, be careful with the imports.

File: basic-application/basic-application-webapp/src/main/java/fr/openwide/core/basicapp/web/application/administration/component/UserGroupMembersPanel.java (hunk 1 above)
Comment (florianlacreuse) [inline basic-application/basic-application-webapp/src/main/java/fr/openwide/core/basicapp/web/application/administration/component/UserGroupMembersPanel.java]: Done.

File: basic-application/basic-application-webapp/src/main/java/fr/openwide/core/basicapp/web/application/common/renderer/ActionRenderers.java (hunk 2 above)
Comment (florianlacreuse) [inline basic-application/basic-application-webapp/src/main/java/fr/openwide/core/basicapp/web/application/common/renderer/ActionRenderers.java]: Done.

File: owsi-core/owsi-core-components/owsi-core-component-wicket-more/src/main/java/fr/openwide/core/wicket/more/markup/html/repeater/data/table/builder/action/CoreActionColumn.java (hunk 3 above)
Comment (florianlacreuse) [inline owsi-core/owsi-core-components/owsi-core-component-wicket-more/src/main/java/fr/openwide/core/wicket/more/markup/html/repeater/data/table/builder/action/CoreActionColumn.java]: Fixed. This snippet was not copied, I'm just so used to write getters. :)

File: owsi-core/owsi-core-components/owsi-core-component-wicket-more/src/main/java/fr/openwide/core/wicket/more/markup/html/repeater/data/table/builder/action/factory/AbstractActionColumnElementFactory.java (hunk 4 above)
Comment (florianlacreuse) [inline owsi-core/owsi-core-components/owsi-core-component-wicket-more/src/main/java/fr/openwide/core/wicket/more/markup/html/repeater/data/table/builder/action/factory/AbstractActionColumnElementFactory.java]: Done.

**Pull request merged**